src/Vault.sol

```diff
@@ -0,0 +1,175 @@
+
+interface IERC20Decimals is IERC20 {
+    function decimals() external view returns (uint);
+}
+
+contract Vault is Ownable {
+
+    ///@dev Vault records collateral deposits of each user
+    mapping (address => uint) public deposits; 
+
+    ///@dev Vault records debt holdings of each user
+    mapping (address => uint) public debts;  
+
+    ///@dev ERC20 interface specifying token contract functions
+    IERC20Decimals public immutable collateral;    
+    IERC20Decimals public immutable debt;  
+
+    ///@dev Asset Pricefeed interface from Chainlink
+    AggregatorV3Interface public immutable priceFeed;   
+    
+    ///@dev Emitted on deposit()
+    ///@param collateralAsset The address of the collateral asset
+    ///@param user The address of the user calling deposit()
+    ///@param collateralAmount The amount of collateral asset deposited
+    event Deposit(address indexed collateralAsset, address indexed user, uint collateralAmount);  
+
+    ///@dev Emitted on borrow()
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling borrow()
+    ///@param debtAmount The amount of debt asset borrowed
+    event Borrow(address indexed debtAsset, address indexed user, uint debtAmount); 
+
+    ///@dev Emitted on repay()
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling repay()
+    ///@param debtAmount The amount of debt asset being repaid
+    event Repay(address indexed debtAsset, address indexed user, uint debtAmount);
+
+    ///@dev Emitted on withdraw()
+    ///@param collateralAsset The address of the debt asset
+    ///@param user The address of the user calling withdraw()
+    ///@param collateralAmount The amount of collateral asset withdrawn
+    event Withdraw(address indexed collateralAsset, address indexed user, uint collateralAmount);  
+
+    ///@dev Emitted on liquidation()
+    ///@param collateralAsset The address of the debt asset
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling withdraw()
+    ///@param debtToCover The amount of debt the liquidator wants to cover
+    ///@param liquidatedCollateralAmount The amount of collateral received by the liquidator
+    event Liquidation(address indexed collateralAsset, address indexed debtAsset, address indexed user, uint debtToCover, uint liquidatedCollateralAmount);
+    
+    ///@notice To allow for margined collateralization; similar to FX brokers
+    ///@dev Fixed-point number; to set collateralization level of debt engine
+    /*Note: collateralLevel is a fixed-point number representation of a decimal (e.g. 0.8), | collateralLevel = range(0, 1e18]
+    *       therefore value supplied has to be order of magnitude smaller than the precision of collateral asset itself.
+    *       Example: 1 DAI (1e18) | 80%: 0.8 DAI (8e17) -> therefore, collateralLevel = 8e17
+    */     
+    uint public collateralLevel; 
```

Comment:
On blockchain it is much more common for the collateralisation ratio to be above 100%. I don't think anyone has managed to build a protocol offering undercollateralised loans yet.

Reply:
Really? Shouldn't this be an easier proposition?

Reply:
Dropped the collateralization ratio aspect.